### ISSUE TITLE
[QST] How is the datasets used for GPTQ? Should i provide chunked tokenized input_ids to oneshot?

### ISSUE THREAD

**Issue by Dingjifeng** [vllm-project/llm-compressor#2233]
I want to use GPTQ & spinquant for Qwen3MoE. However I don't know how to provide the correct datasets. Below is my code:
```
import torch
from transformers import AutoModelForCausalLM, AutoTokenizer
from datasets import load_dataset, Dataset  # <--- 1. 必须引入 Dataset
from llmcompressor import oneshot
from llmcompressor.modifiers.quantization import GPTQModifier
from llmcompressor.modifiers.transform import SpinQuantModifier
import itertools

# 1. 模型设置
MODEL_ID = "/root/workspace/models/qwen3-moe-235b"

model = AutoModelForCausalLM.from_pretrained(
    MODEL_ID, 
    device_map="auto", 
    torch_dtype="auto",
    trust_remote_code=True 
)
tokenizer = AutoTokenizer.from_pretrained(MODEL_ID, trust_remote_code=True)

# 2. 数据集处理
SEQ_LEN = 4096
NUM_CALIBRATION_SAMPLES = 256

ds = load_dataset("EleutherAI/wikitext_document_level", "wikitext-2-raw-v1", split="train")

def preprocess_and_concat(dataset):
    tokenized_list = []
    
    # 自动检测列名，防止 key error
    # wikitext_document_level 通常用 "page"，标准 wikitext 用 "text"
    text_column = "page" if "page" in dataset.column_names else "text"
    print(f"Detected text column: {text_column}")

    for sample in dataset:
        text = sample[text_column]
        if len(text) > 0:
            tokenized_list.append(tokenizer(text, add_special_tokens=False)["input_ids"])
    
    # 展平
    concat_ids = list(itertools.chain.from_iterable(tokenized_list))
    
    # 切分
    total_tokens = len(concat_ids)
    num_batches = total_tokens // SEQ_LEN
    limit = min(num_batches, NUM_CALIBRATION_SAMPLES)
    
    input_ids = [concat_ids[i * SEQ_LEN : (i + 1) * SEQ_LEN] for i in range(limit)]
    
    # ---------------------------------------------------------
    # 核心修复：这里必须返回 Dataset 对象，而不是 list
    # ---------------------------------------------------------
    return Dataset.from_list([{"input_ids": ids} for ids in input_ids])

# 生成校准数据 (现在是 Dataset 对象了)
ds_processed = preprocess_and_concat(ds)

print(f"Prepared {len(ds_processed)} samples with sequence length {SEQ_LEN}")
# 验证一下类型，确保它是 Dataset
print(f"Dataset type: {type(ds_processed)}") 

# 3. 配置 Recipe
recipe = [
    SpinQuantModifier(
        rotations=["R1", "R2", "R4"],
        transform_block_size=128,
        transform_type="hadamard",
    ),
    GPTQModifier(
        targets="Linear", 
        scheme="FP8", 
        dampening_frac=0.1, 
        block_size=128, 
        ignore=[
            "lm_head", 
            "re:.*mlp.gate$",
        ]
    ),
]

# 4. 执行 OneShot
oneshot(
    model=model, 
    dataset=ds_processed, # 这里现在传入的是正确的 Dataset 对象
    recipe=recipe, 
    num_calibration_samples=NUM_CALIBRATION_SAMPLES, 
    output_dir="./qwen3-moe-235b-quantized-fp8",
    save_compressed=True,
    trust_remote_code_model=True
)
```
I used chunked inputs_id cause I don't know if 

```
ds = load_dataset("EleutherAI/wikitext_document_level", "wikitext-2-raw-v1", split="train")


def tokenize(sample):
    return tokenizer(
        sample["page"],  
        add_special_tokens=False,
    )

ds = ds.map(tokenize, remove_columns=ds.column_names)
```
This simply works? Can someone help me?

**Comment by kylesayrs** [comment]
Hey @Dingjifeng 

For a dataset like `EleutherAI/wikitext_document_level`, I would process as something like this:

```python3
def tokenize(sample):
    return tokenizer(
        sample["page"],  
        add_special_tokens=False,
        max_sequence_length=2048
        truncate=True,
        padding=False,
    )
```

This makes sure that you never have a sample which starts in the middle of a sentence. You'll only need ~512 samples to get a quantization.

**Comment by kylesayrs** [comment]
A couple other things to note:

Please load your model on CPU (`oneshot` will handle the moving to GPU
```python3
model = AutoModelForCausalLM.from_pretrained(
    MODEL_ID, 
    # device_map="auto", 
    torch_dtype="auto",
    trust_remote_code=True 
)
```

Your recipe is likely a bit overcomplicated. Modifiers like `GPTQModifier` will increase your runtime. FP8 is a near loss-less format, so using the `QuantizationModifier` which is faster and less memory intensive will yield just as high accuracy recovery.
```python3
recipe = [
    #GPTQModifier(
    QuantizationModifier(
        targets="Linear", 
        scheme="FP8",
        ignore=[
            "lm_head", 
            "re:.*mlp.gate$",
        ]
    ),
]
```

**Comment by Dingjifeng** [comment]
Hi

> A couple other things to note:
> 
> Please load your model on CPU (`oneshot` will handle the moving to GPU
> 
> model = AutoModelForCausalLM.from_pretrained(
>     MODEL_ID, 
>     # device_map="auto", 
>     torch_dtype="auto",
>     trust_remote_code=True 
> )
> Your recipe is likely a bit overcomplicated. Modifiers like `GPTQModifier` will increase your runtime. FP8 is a near loss-less format, so using the `QuantizationModifier` which is faster and less memory intensive will yield just as high accuracy recovery.
> 
> recipe = [
>     #GPTQModifier(
>     QuantizationModifier(
>         targets="Linear", 
>         scheme="FP8",
>         ignore=[
>             "lm_head", 
>             "re:.*mlp.gate$",
>         ]
>     ),
> ]

Thanks for the quick response. Still have few question about applying SpinQuant, If I add SpinQuant in recipe, should I  manually rotate the activation? Is it the same way to do with SpinQuant's repo? by adding hook before "forward" the activation? How can i know the hadamard matrix is the same? also, i noticed when i run the original code of mine, it gives calibrating on Spinquant, these seems wired? since I expect only do rotate the weight and move forward to GPTQ directly.  @kylesayrs Thanks for your reply
